# Main entrypoint of the workflow. 
# Please follow the best practices: 
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there. 


# Main Snakefile

# set config file where you can specify your inputs

configfile: "config/config.yaml"

# Include the hifi_prep rule
include: "rules/hifi_prep.smk"

# Include the kmc rule
include: "rules/kmc.smk"

# Include the genomescope2 rule
include: "rules/genomescope2.smk"

# Include the oatk rule
include: "rules/oatk.smk"

# Include the minimap2 rule
include: "rules/minimap2.smk"

# Include the samtools rule
include: "rules/samtools.smk"

# Include the seqkit rule
include: "rules/seqkit.smk"

# Include the seqkit rule
include: "rules/hifiasm.smk"

# Include the purge_dups rule
include: "rules/purge_dups.smk"

# Define the rule all, which is the default rule that Snakemake runs when no specific rule or target is specified
rule all:
    input:
        "results/reads/hifi/hifi.fastq.gz", # output of hifi_prep rule
        "results/kmc/out.hist", # output of kmc rule
        "results/genomescope", # output directory of genomescope
        "results/oatk/oatk.asm.mito.ctg.fasta", # mitochondrion fasta from oatk
        "results/oatk/oatk.asm.pltd.ctg.fasta", # chloroplast fasta from oatk
        "results/minimap2/aln.mito.sam", # minimap2 alignment for mitochondrion
        "results/minimap2/aln.pltd.sam", # minimap2 alignment for chloroplast
        "results/samtools/organelles.reads.txt", # organelles reads to exclude, result of samtools
        "results/reads/hifi/filtered_orgs_reads.fastq.gz", # hifi excluding organelles reads
        "results/assemblies/hifiasm/hifiasm.asm.bp.p_ctg.gfa", # draft hifiasm assembly.gfa
        "results/assemblies/hifiasm/hifiasm.asm.bp.p_ctg.fa", # draft hifiasm assembly.fa
        "results/assemblies/purge_dups/hifiasm.asm.purged.fa" # purged assembly with purge_dups